### PR TITLE
Add Task.requestCustomJSONEncodable

### LINKED ISSUE #1442: How to customize JSONEncoder for Task.requestJSONEncodable?
I need a way to pass custom DateFormatter to dateEncodingStrategy, but it seems now it is impossible.

Moya version is 10.0.0.

### PR DESCRIPTION
**Pull request by evgeny-sureev**
Fixes #1442

### CODE REVIEW

File: Tests/EndpointSpec.swift [hunk 1]
```diff
@@ -175,6 +175,45 @@ final class EndpointSpec: QuickSpec {
                 }
             }
 
+            context("when task is .requestCustomJSONEncodable") {
```
Comment (SD10) [inline Tests/EndpointSpec.swift]: Maybe we can add test case for when encoding fails?

**Comment by SD10** [comment]
Hey @evgeny-sureev, thanks for doing this 👍 I think this PR is better targeted at development because the addition of a new `Task` case is an API breaking change. However, I see that `development` needs to be updated to reflect `master`

**Comment by evgeny-sureev** [comment]
First SwiftLint error is Cyclomatic Complexity Violation, it is unavoidable in switches for enums with many cases.

**Comment by AndrewSB** [comment]
@evgeny-sureev can associated values in an enum case be given defaults? If so, maybe it would be nice to add an encoder property to the `requestJSONEncodable` case instead of creating a new one. For people who don't want to change the default value, they need not use it

**Comment by evgeny-sureev** [comment]
This would be best variant, but Swift doesn't support defaults for enum associated values yet.

**Comment by Istered** [comment]
You can use static functions to "mimic" cases with default values.
Something like this:
```swift
public enum Task {
   ....
    case requestCustomJSONEncodable(Encodable, encoder: JSONEncoder)

    public static func requestJSONEncodable(_ encodable: Encodable, encoder: JSONEncoder = JSONEncoder()) -> Task {
        return .requestCustomJSONEncodable(encodable, encoder: encoder)
    }
   ....
}
```

Or even skip the second parameter:
```swift
public static func requestJSONEncodable(_ encodable: Encodable) -> Task {
    return .requestCustomJSONEncodable(encodable, encoder: JSONEncoder())
}
```

**Comment by sunshinejr** [comment]
This is really smart @Istered! Doesn't compiler complain about it, though? 🤔

**Comment by BasThomas** [comment]
That should just work - we're actually doing something similar already [here](https://github.com/Moya/Moya/blob/6a37ac5d45dded0ffd008b76a41eb31c5dcf33b1/Sources/Moya/MoyaProvider.swift#L147-L163).

**Comment by Istered** [comment]
@sunshinejr It will just work as long as functions and cases have different names.

**Comment by sunshinejr** [comment]
Oh, so even though the parameter list is different, it still won't work right @Istered? Bummer. Then we probably can't make it pretty without duplicating cases, right?

**Comment by Istered** [comment]
@sunshinejr Yes, it won't work even though parameter list is different, but still there's only one case you need to handle with switch.

So instead of this:
```swift
switch task {
    ....
     /// These are two different cases to handle Encodable in this PR
    case let .requestJSONEncodable(encodable):
     ///

    case let .requestCustomJSONEncodable(encodable, encoder):
     ///
    ....
}
```
You would write only this:
```swift
switch task {
    ....
     /// 
    case let .requestCustomJSONEncodable(encodable, encoder):
     ///
    ....
}
```

**Comment by evgeny-sureev** [comment]
@Istered done.

**Comment by SD10** [comment]
I'm not sure I agree with this approach. We're avoiding adding another enum case (technically replacing an existing one) by adding a static method. 

I don't think there is anything to gain here, we're simply moving the definition from the `case` level to the `method` level and making the API less intuitive.

I also really dislike that the default name of this task is now `requestCustomJSONEncodable`

**Comment by sunshinejr** [comment]
I agree with @SD10. This is why I said I don't see much advantages in that approach with our current situation.

**Comment by Istered** [comment]
@SD10 @sunshinejr 
Maybe it's just me, but I don't see any problems with that since implicit member expressions work with static functions the same way as with enum cases.

What really confuses me are these two lines doing absolutely the same:
```swift
return .requestJSONEncodable(...)
return .requestCustomJSONEncodable(..., JSONEncoder())
```

But I agree with you, guys, it would be better to have only one case. However, the problem is that you can't specify default value for cases as it was mentioned before and also it would be more or less breaking change, right?

**Comment by SD10** [comment]
@Istered I'm not worried about breaking the API, breaking changes are inevitable and this requires a very simple migration.

```Swift
// These may "do" the same thing but the way they are defined have implications
return .requestJSONEncodable(...)
return .requestCustomJSONEncodable(..., JSONEncoder())
```

What bothers me is there is no value in defining `requestJSONEncodable(Encodable, encoder: JSONEncoder)` as a static method.

- It's true that there will be 1 less enum case but the code footprint is no smaller, it actually takes more code to define a method.

-  It makes the API non-intuitive, `Task.requestJSONEncodable` looks like an enum case to the user but it is really a method that returns a `Task.requestCustomJSONEncodable`.

- We no longer get the added benefit of exhaustiveness for having this defined as an enum case.

I think we should solve this problem 1 of two ways:

a) We make one case `requestJSONEncodable(Encodable)` and the user just has to use a boilerplate `JSONEncoder()` or create their own convenience property.

b) We have two cases: `requestJSONEncodable(Encodable, encoder: JSONEncoder)` and `requestCustomJSONEncodable(Encodable, encoder: JSONEncoder)`

**Comment by SD10** [comment]
@evgeny-sureev Can we revert the last commit and move forward on merging this? That would be in line with option (B) I mentioned in my previous comment. Unless any @Moya/contributors disagree with me of course

**Comment by AndrewSB** [comment]
@SD10: tbh I prefer (a), it seems more complete a solution

**Comment by evgeny-sureev** [comment]
I think I was too fast with last commit. Feel free to revert it.

**Comment by sunshinejr** [comment]
We have already a method that requires only to pass `Encodable` and if you don't worry about your `JSONEncoder` (which is probably the case most of the times), it is tedious to pass an empty instance to the method all the time (we have it already e.g. request with/without parameters/encoding etc.). Thus, I'm more keen to just add another case to our `Task` type.

Thoughts @evgeny-sureev @SD10 @AndrewSB?

**Comment by SD10** [comment]
@sunshinejr I agree that adding another case best matches the existing request/parameter encoding API. It doesn’t make sense to treat encodable requests any different. If we’re not going to require users to specify parameter encoding on every request then we shouldn’t require them to pass a JSON decoder (which is more likely to remain constant).

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by evgeny-sureev** [comment]
@sunshinejr I think adding new case to `Task`, as I did earlier, is better.

**Comment by sunshinejr** [comment]
Cool! @evgeny-sureev can we go back to two cases then? :)

**Comment by evgeny-sureev** [comment]
Ok, now we have two cases.

**Comment by SD10** [comment]
@evgeny-sureev Thanks 👍 It looks good to me. If you could just two do more things:
1) Fix the SwiftLint warning
2) Add a CHANGELOG entry under **Added**, this is a breaking change

**Comment by evgeny-sureev** [comment]
Do you have an idea how to fix cyclomatic complexity warning? Other warnings are not a result of this pull request.

**Comment by SD10** [comment]
@evgeny-sureev I don't see a good way to fix that one. When I made my comment Moya had not updated the comment from the results of CI. The build was failing so I restarted it

**Comment by evgeny-sureev** [comment]
Maybe add '// swiftlint:disable:next cyclomatic_complexity'?

**Comment by SD10** [comment]
@evgeny-sureev Sure that would be fine by me. It looks like there are 39 failing tests on CI

File: Tests/EndpointSpec.swift [hunk 2]
```diff
@@ -286,6 +325,26 @@ final class EndpointSpec: QuickSpec {
                 }
             }
 
+            context("when JSONEncoder set with incorrect parameters") {
```
Comment (sunshinejr) [inline Tests/EndpointSpec.swift]: nice one 👌

**Comment by ashfurrow** [comment]
Thanks a lot for contributing to Moya! I've invited you to join the Moya GitHub organization – no pressure to accept! If you'd like more information on what that means, check out our [contributor guidelines](https://github.com/Moya/contributors) and feel free to reach out with any questions.

**Comment by AndrewSB** [comment]
Uggh, I don't particularly like the trade-off we made, but I understand why we did. Thank you for your work and patience @evgeny-sureev 👍

**Comment by Lang-le** [comment]
@evgeny-sureev How can i send url parameters with .requestCustomJSONEncodable? Please help me!

**Comment by SD10** [comment]
@Lang-le Your question is better suited to opening a new GitHub issue

**Comment by Lang-le** [comment]
@SD10 Thank you! I will open it!

**Pull request merged**